### PR TITLE
optimize slice_input_producer

### PR DESCRIPTION
**Pull request by ppwwyyxx**
When shuffle=False, input_slice_producer can be implemented directly with FIFOQueue.
The original implementation using range and array indexing is slower.

I could pass the unit test `input_test.py`. But please correct me if I could use the name and `set_shape` in a better way (i'm not totally sure).

I simply run it on my old laptop on CPU and I saw `input_slice_producer + batch` operations (which is a no-op) introduces no overhead, while with the original implementation I see obvious overhead.

### CODE REVIEW

**Comment by mrry** [comment]
One issue with this change is that it makes multiple copies of the input data (by default 32), which could have an unfortunate effect on memory usage. Perhaps a cleaner optimization would be to produce intervals (rather than ranges), which could be used to `tf.slice()` the appropriate elements from the input tensor(s), rather than `tf.gather()` them?

**Comment by ppwwyyxx** [comment]
Thanks. I'm not sure what do you mean. What's the different between __interval__ and __range__?

And also, why do you think a slice operation would be faster than gather? It looks like they are doing the same thing since we only need to produce one element at a time.

**Comment by mrry** [comment]
Sorry for the delay in getting back to you! Looking at your PR more closely, it seems like I misunderstood the nature of your change (and the nature of `slice_input_producer()`, which I had mistakenly assumed produces batches rather than single elements...). I'm still a bit unclear about why your change makes things more efficient: as far as I can tell, the change will end up copying more data out of `input_tensor` on each iteration, which is potentially wasteful. Do you have details of a benchmark that shows the improvement?

Looking at the implementation, it also looks like the `array_ops.gather()` is the only thing that your change avoids, but at the expense of copying the input tensor into the queue repeatedly, which could potentially lead to a performance regression. It seems like you could get almost all of the benefit by replacing the `array_ops.gather()` with an `array_ops.slice()`, which in many cases could avoid the copy altogether.

Does that make sense?

**Comment by ppwwyyxx** [comment]
```python
import tensorflow as tf
import numpy as np
import time
import sys
import cv2

def get_output(img):
    img = tf.train.slice_input_producer([img], shuffle=False, capacity=32)
    img = tf.train.batch(img, 128)

    img = tf.reshape(img, [-1, 28 * 28 * 3])
    W = tf.get_variable('W', [28 * 28 * 3, 10],
                        initializer=tf.truncated_normal_initializer(stddev=0.1))
    out = tf.matmul(img, W)
    return tf.sigmoid(out)

G = tf.Graph()
with G.as_default():
    FAKE = np.random.rand(28, 28, 3)
    img = tf.constant(FAKE, dtype=tf.float32, shape=[28, 28, 3])
    img = tf.train.batch([img], 128)
    output = get_output(img)

    sess = tf.Session()
    sess.run(tf.initialize_all_variables())

    coord = tf.train.Coordinator()
    tf.train.start_queue_runners(
        sess=sess, coord=coord, daemon=True, start=True)

    start = time.time()
    for k in range(5):
        sess.run(output)
        print k
    print time.time() - start, " seconds"
    coord.request_stop()
```
I made a script for testing. On my laptop without GPU, I see the following numbers:
+ Run with 0.6.0 release: 16s.
+ With my patch: 0.25s
+ Without `slice + batch` (delete the first two lines in `get_output`): 0.14s
And it surprises me that a `slice + batch` can slow down the run time so much. Last time I tested (with real data and models) the difference wasn't that much.
I'm also interested in why that happens. Maybe there are ways to optimize this script that I didn't notice.

Also, I noticed that if I passed a constant variable of shape 128x28x28x3 to `get_output` instead, there is no performance issue anymore.

**Comment by mrry** [comment]
I think that there's a better way to solve your problem:

```python
def get_output(img):
    img = tf.train.limit_epochs([img], num_epochs=None)  # Produce infinitely many times.
    img = tf.train.batch(img, 128, enqueue_many=True)  # Enqueues a batch at a time.

    img = tf.reshape(img, [-1, 28 * 28 * 3])
    W = tf.get_variable('W', [28 * 28 * 3, 10],
                        initializer=tf.truncated_normal_initializer(stddev=0.1))
    out = tf.matmul(img, W)
    return tf.sigmoid(out)
```

On my workstation, this modified version of your benchmark takes between 0.05 and 0.08 seconds.

**Comment by ppwwyyxx** [comment]
In this case of course, this script is just made to test the speed of `input_slice_producer.`
If for example I want to apply some image processing functions (random brightness, etc) on each image in a batch, then I may need to use slice producer to take single image from a batch.

**Comment by vrv** [comment]
What's the status of this?

**Comment by vrv** [comment]
Closing due to inactivity.
